### DISCUSSION TITLE
PBKDF2 - compatibility between Elixir and Rust/Stalwart

### DISCUSSION THREAD

**Discussion by lulu-2021** [stalwartlabs/jmap-server#57] [Q&A]
Hi there - am the process of writing a web based admin app in Elixir & Phoenix for User admin - and I have implemented the secret hashing using this library --> https://hex.pm/packages/pbkdf2_elixir -- it's using the standard pbkdf2 algorithm and the salting is done by default with SHA-512 -- when authenticating with this user stalwart spits out "Invalid password hash" -- just to verify that there was not something else going on - I updated the secret using "openssl passwd -6" to generate a SHA512 hash with the same password and I could authenticate ok.... am left wondering whether the Rust implementation of PBKDF2 is not compatible with the ELIXIR (Erlang) version ? Any idea why that would not be compatible..? I also tried this library (https://github.com/thomas9911/password-rs) - as it uses a rustler based nif - and it causes the same log / error -- your feedback would be very much appreciated - thanks

**Comment by mdecimus** [comment]
Hi,

Stalwart uses RustCrypto's [pbkdf2](https://crates.io/crates/pbkdf2) library. Make sure that the password hashes generated by your app begin with `$pbkdf2`. Hope it helps!

**Reply by lulu-2021** [reply]
hi, they do - which is why I was surprised... could it be to do with the "phc-sf-spec" - i.e. the PHC string format ? (Salt, version, memory time, paralellism) -- the specification --> "Parameter String" "Salt String" "Hash String" -- ?

**Reply by lulu-2021** [reply]
Hi there - I tested to specify the "SHA" version i.e. "pbkdf2-sha512" in the algorithm - which changes that in the secret and this was accepted by Stalwart - so maybe it would be a good idea to mention this in the documentation for other people who might have the same issue. I will let you know when I am done with the "admin interface" -- for now it's just for me as a SysAdmin to make my life easier with the implementation of Stalwart for my clients -- I use it myself as well - wanted to congratulate you on the great job you have done. Finally there is a product that allows the client to upload their public PGP key and all email is stored encrypted and inaccessible to the sysadmin... that is awesome in this day and age of "surveillance on steroids".... this product should be given a "Liberty Award" - Thanks

**Reply by mdecimus** [reply]
Thanks! Will take note of that. It is strange that the pbkdf2 crypto library does not handle this automatically.
